tests/ui/implied-bounds/gluon_salsa.rs

```diff
@@ -18,12 +17,17 @@ pub struct QueryTable<'me, Q, DB> {
 }
 
 impl<'me, Q> QueryTable<'me, Q, <Q as QueryBase>::Db>
-// projection is important
-//   ^^^ removing 'me (and in QueryTable) gives a different error
 where
-    Q: for<'f> AsyncQueryFunction<'f>,
+    Q: for<'f> QueryFunction<'f>,
 {
+    // When computing the implied bounds here we normalize `<Q as QueryBase>::Db` to
+    // `<Self as QueryFunction<'?x>>::SendDb`. This is not usable as an
+    // implied bound due to the unconstrained lifetime.
+    //
+    // The `<Q as AsyncQueryFunction<'?y>>::SendDb`: 'a` bound then results in an error
+    // during typeck.
     pub fn get_async<'a>(&'a mut self) {
```

Comment:
this one is interesting, have to figure out why this test previously compiled

Reply:
this previously compiled as in `match_poly_trait_ref` we used `sup` which ends up using `Sub` with lhs and rhs flipped. This results in the infer var in the normalized self type to not be the root infer var. We then use this self type for the implied bounds, canonicalization of which does opportunistically resolve regions, so the resulting implied bound uses the semantically equal, but not structurally equal root infer var.

Computing the verify bounds for the alias outlives then does a structural equality check which now fails, as we never opportunistically resolve regions in outlives bounds.

The following diff fixes this:
```diff
diff --git a/compiler/rustc_infer/src/infer/outlives/obligations.rs b/compiler/rustc_infer/src/infer/outlives/obligations.rs
index c0a99e5cc41..a17b099bf4f 100644
--- a/compiler/rustc_infer/src/infer/outlives/obligations.rs
+++ b/compiler/rustc_infer/src/infer/outlives/obligations.rs
@@ -62,6 +62,7 @@
 use crate::infer::outlives::components::{push_outlives_components, Component};
 use crate::infer::outlives::env::RegionBoundPairs;
 use crate::infer::outlives::verify::VerifyBoundCx;
+use crate::infer::resolve::OpportunisticRegionResolver;
 use crate::infer::{
     self, GenericKind, InferCtxt, RegionObligation, SubregionOrigin, UndoLog, VerifyBound,
 };
@@ -71,6 +72,7 @@
 use rustc_middle::traits::query::NoSolution;
 use rustc_middle::ty::{self, GenericArgsRef, Region, Ty, TyCtxt, TypeVisitableExt};
 use rustc_middle::ty::{GenericArgKind, PolyTypeOutlivesPredicate};
+use rustc_middle::ty::TypeFoldable;
 use rustc_span::DUMMY_SP;
 use smallvec::smallvec;
 
@@ -177,6 +179,7 @@ pub fn process_registered_region_obligations(
                         .no_bound_vars()
                         .expect("started with no bound vars, should end with no bound vars");
 
+                let (sup_type, sub_region) = (sup_type, sub_region).fold_with(&mut OpportunisticRegionResolver::new(self));
                 debug!(?sup_type, ?sub_region, ?origin);
 
                 let outlives = &mut TypeOutlives::new(
```

Reply:
Does that diff result in any other differences?

Reply:
See the diagnostic changes in #121743